### PR TITLE
[ListViewDataSource] Simplify API: constructor accepts rows and provides default rowHasChanged

### PR DESCRIPTION
**Pull request by brentvatne**
As per #1595

```javascript
getInitialState() {
  var ds = new ListView.DataSource({rowHasChanged: (r1, r2) => true});
  return {
    dataSource: ds.cloneWithRows(['row 1', 'row 2']),
  };
},

render() {
  return (
    <ListView
      dataSource={this.state.dataSource}
      renderRow={(rowData) => <Text>{rowData}</Text>}
    />
  );
},
```

becomes

```javascript
getInitialState() {
  return {
    dataSource: new ListView.DataSource({rows: ['row 1', 'row 2']});
  };
},

render() {
  return (
    <ListView
      dataSource={this.state.dataSource}
      renderRow={(rowData) => <Text>{rowData}</Text>}
    />
  );
},
```

### CODE REVIEW

File: Libraries/CustomComponents/ListView/ListViewDataSource.js [hunk 1]
```diff
@@ -112,31 +125,57 @@ class ListViewDataSource {
...
+      } else {
+        this.rowIdentities = [];
+        this.sectionIdentities.forEach((sectionID) => {
+          this.rowIdentities.push(Object.keys(this._dataBlob[sectionID]));
+        });
+      }
+      this._cachedRowCount = countRows(this.rowIdentities);
+
+      // Dirty all rows and sections
+      this._calculateDirtyArrays([], [], []);
```
Comment (brentvatne) [inline Libraries/CustomComponents/ListView/ListViewDataSource.js]: TODO: have the constructor accept `prevData` param, so we don't do extra work when cloning

**Comment by paramaggarwal** [comment]
This does make the API simpler. Only concern is that new folks might not realise that they can optimise by using `rowHasChanged`. This is different from before where it was explicitly declared in the beginning and hence known beforehand.

While we are at it we can change `rowHasChanged` to `rowShouldUpdate` to be like React's `componentShouldUpdate` method.

**Comment by ide** [comment]
> While we are at it, we can change rowHasChanged to rowShouldUpdate to be like React's componentShouldUpdate method.

I like this. Probably makes more sense as a separate PR. We'll want to keep the old method - and have it print a helpful warning - around for a release or two so that people have time to migrate.

**Comment by sahrens** [comment]
Yeah, this interface definitely needs to be better.  I'm wondering if maybe we should just make a completely new class, maybe `ListView.SimpleDataSource`, instead of cluttering up the already cluttered `ListViewDataSource`?

Some other changes I'd like to make are to make the data blob headers structured data, so you could do something like this:

  dataSourceWithSections: new ListView.SimpleDataSource({sections: [
    {header: 'h1', rows: ['row 1a', 'row 1b']},
    {header: 'h2', rows: ['row 2a']}
  ]}),
  dataSourceWithoutSections: new ListView.SimpleDataSource({rows: [
    'row 1',
    'row 2',
  ]}),

what do you think?

re: rowShouldUpdate, I think we should actually probably ditch that completely along with the `StaticRenderer`s and rely entirely on `componentShouldUpdate` of the actual row components, rather than adding additional layers of magic.  We could also potentially add some timing code in render and it takes a long time and the components don't have `shouldComponentUpdate` implemented, we could warn.

**Comment by aleclarson** [comment]
:+1:

**Comment by paramaggarwal** [comment]
@brentvatne Missed this PR today. ListView API currently needs some setup to get started - this PR simplifies that. Great comments from @sahrens on how to take this forward.

**Comment by brentvatne** [comment]
Good point @paramaggarwal - I totally forgot about this :p

**Comment by mkonicek** [comment]
Hi there! This issue is being closed because it has been inactive for a while.

But don't worry, it will live on with ProductPains! Check out its new home: https://productpains.com/post/react-native/listviewdatasource-simplify-api-constructor-accepts-rows-and-provides-default-rowhaschanged

ProductPains helps the community prioritize the most important issues thanks to its voting feature.
It is easy to use - just login with GitHub.

Also, if this issue is a bug, please consider sending a PR with a fix. We rely on the community to provide
bugfixes as the core team is focused on performance.

**Pull request closed without merging**